subset/SSet.cpp

```diff
@@ -0,0 +1,105 @@
+#include "SSet.h"
+#include <algorithm>
+#include <iterator>
+
+bool SSet::checkSumOptimized(DataType expected) {
+  positive_sum = 0;
+  negative_sum = 0;
+  for (auto n : set_) {
+    if (n > 0) positive_sum += n;
+    if (n < 0) negative_sum += n;
+  }
+  return set_.empty() ? false
+                      : checkSumOptimizedRecursive(set_.size() - 1, expected);
+}
+
+bool SSet::checkSumOptimizedRecursive(int i, DataType expected) {
+  if (i == 0) return set_[i] == expected;
+  if ((expected > positive_sum) || (expected < negative_sum)) return false;
+  return checkSumOptimizedRecursive(i - 1, expected) || (set_[i] == expected) ||
+         checkSumOptimizedRecursive(i - 1, expected - set_[i]);
+}
+
+std::vector<SSet::DataType> SSet::subsetSums(std::vector<DataType>* set_p) {
```

Comment:
Quite strange function. Take pointer to an vector and copy vector itself. Then use this copy for reading elements. This is redundant